File: src/main/java/net/flyingfishflash/loremlist/domain/exceptions/DomainException.java
package net.flyingfishflash.loremlist.domain.exceptions;

import com.fasterxml.jackson.databind.JsonNode;
import java.net.URI;
import java.util.Map;
import net.flyingfishflash.loremlist.core.exceptions.CoreException;
import org.springframework.http.HttpStatus;

public class DomainException extends CoreException {

  public static final String DEFAULT_TITLE = "DomainException";
  public static final String DEFAULT_MESSAGE = DEFAULT_TITLE;

  public DomainException(
      Throwable cause,
      HttpStatus httpStatus,
      String message,
      String responseMessage,
      String title,
      URI type,
      Map<String, JsonNode> supplemental) {
    super(
        cause,
        httpStatus,
        message != null ? message : DEFAULT_MESSAGE,
        responseMessage,
        title != null ? title : DEFAULT_TITLE,
        type,
        supplemental);
  }

  public static Builder<?> builder() {
    return new Builder<>();
  }

  public static class Builder<B extends Builder<B>> extends CoreException.Builder<B> {
    @Override
    public DomainException build() {
      return new DomainException(
          cause, httpStatus, message, responseMessage, title, type, supplemental);
    }
  }
}

File: src/main/kotlin/net/flyingfishflash/loremlist/domain/exceptions/DomainException.kt
package net.flyingfishflash.loremlist.domain.exceptions

import kotlinx.serialization.json.JsonElement
import net.flyingfishflash.loremlist.core.exceptions.CoreException
import org.springframework.http.HttpStatus
import java.net.URI

open class DomainException(
  cause: Throwable? = null,
  httpStatus: HttpStatus? = null,
  message: String? = null,
  responseMessage: String? = null,
  title: String? = null,
  type: URI? = null,
  supplemental: Map<String, JsonElement>? = null,
) : CoreException(
  cause = cause,
  httpStatus = httpStatus,
  message = message ?: DEFAULT_MESSAGE,
  responseMessage = responseMessage,
  title = title ?: DEFAULT_TITLE,
  type = type,
  supplemental = supplemental,
) {
  companion object {
    const val DEFAULT_TITLE = "DomainException"
    const val DEFAULT_MESSAGE = DEFAULT_TITLE
  }
}
